### DISCUSSION TITLE
Navigation between Pages/UserControls guide

### DISCUSSION THREAD

**Discussion by wyndn** [AvaloniaUI/Avalonia#11807] [General]
I've been using ReactiveUI to navigate between pages, but I want to transition to MVVM Community Toolkit. There aren't any guides on how to go about doing this, and all the previous recipes for WPF navigation involves CommandManager, which is not implemented in ICommand of Avalonia. 

Can someone point me on where to look concerning this. 

(Also, updating to Reactive 11.0.rcc1 is kinda broken. ViewLocator seems to have disappeared. Can't find a changelog of these on Reactive or Avalonia docs.)

**Comment by timunie** [comment]
ViewLocator is not in the template anymore as it used reflection which breaks trimming and AOT most of the time. However, this class can be easily implemented back by adding a class that implement `IDataTemplate`. Example: https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/Routing/BasicViewLocatorSample#what-is-a-viewlocator

**Reply by wyndn** [reply]
Behold! It works! Thanks a bunch. Guess I'll stick to Reactive UI for a bit longer. If I can query another little thing: Is it possible to use routing and navigation using MVVM Community Toolkit? I've checked the AngleSix YouTube Playlist, but it's really move between pages n such.

**Reply by timunie** [reply]
>  Is it possible to use routing and navigation using MVVM Community Toolkit?

Not needed for `ViewLocator`. Also you can use RxUI side by side with the CT-MVVM package. I do it as well. Only the WhenActivated-stuff is missing, but I don't need that often. And if, there are alternatives. `ReactiveCommand`s can be replaced by `RelayCommand`

**Reply by wyndn** [reply]
Aha. That's a cool solution. Never considered that angle myself. Taking best of both worlds. It streamlines prototyping views and events so much! Thanks again, Timunie.

**Reply by tangtemas** [reply]
https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/Routing/BasicViewLocatorSample#what-is-a-viewlocator

Watting for covert to communicate mvvm toolkit 💯

**Reply by tangtemas** [reply]
💯Can you give me the source code communicate mvvm toolkit?

**Comment by timahrentlov** [comment]
>  Only the WhenActivated-stuff is missing, but I don't need that often. And if, there are alternatives. 
@timunie May I ask, what alternatives you use? Are you rolling your own in a IDataTemplate ?

**Reply by timunie** [reply]
Yeah implement own IDataTemplate is powerful. Beside that, in your Usercontrol you can listen to OnAttachedToVisualTree. Then you can do init stuff there. Last but not least check out IMessanger API  of CT. That way you can correspondence between loose classes.

**Reply by timahrentlov** [reply]
Thank you for the tip of OnAttachedToVisualTree. 

I'll be sure to look into IMessenger. I know it is implemented with weak references. Back in my XF days I got burned by their implementation, which was one of the reasons I got to know about Subjects, Observables, etc. So thank you to XF team for that :)

**Comment by mysteryx93** [comment]
[MvvmDialogs](https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs) allows navigation for mobile devices in the same way you'd show dialogs in Windows, calling `IDialogService.Show` and `IDialogService.ShowDialog` from the ViewModel. There's also the option of using navigation mode also on desktop.

Not sure what your needs are, but there might be something for you there.
